### PR TITLE
Terraform module for Openstack deployments of Lattice

### PR DESCRIPTION
**Pull request by metahertz**
Openstack Terraform module for Lattice.
Used just like the existing AWS/DigitalOcean modules.

Tested with Coordinator and two cell's.
Each instance is given a floating IP to allow remote-exec configuration/installation (as per AWS module).

Test pre-merge by following the readme but pulling the base `lattice.openstack.tf`file from the following location:

https://raw.githubusercontent.com/matjohn2/lattice/openstack-terraform/terraform/openstack/example/lattice.openstack.tf

CLI tests passing against demo deployment.
image

Comments and confirmations welcome!

### CODE REVIEW

File: terraform/openstack/README.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
+# Work in progress!
+  Dont expect this to work yet!
```
Comment (ajlanghorn) [inline terraform/openstack/README.md]: s/Dont/Don't.
Because pedant.

Comment (matjohn2) [inline terraform/openstack/README.md]: Shouldn't you be doing some real work ;).
Actually, do appreciate that, I need to take it out of the readme, missed it.

**Comment by jbayer** [comment]
@matjohn2 we are working on getting a CI environment with openstack so we can validate and accept this PR. we're working on it, thanks for the PR!

**Comment by davebotelho** [comment]
I keep getting the following error.

* Error creating OpenStack keypair: Expected HTTP response code [200] when accessing [POST http://vc2cmmka033430n.fmr.com:8774/v2/f2c27ae1651849e98f80b60231b752ef/os-keypairs], but got 400 instead
{"badRequest": {"message": "Keypair data is invalid: failed to generate fingerprint", "code": 400}}


I'm thinking it has something to do with my config file entries below.  I've created a keypair in my openstack tenant and saved the private key. That is the key used to ssh to each of the servers, so what am I missing?

    # OpenStack Tenant Name
    openstack_tenant_name = "my_dev"

    # SSH Key Name
    openstack_key_name = "my_keypair"

    # SSH Public Key to Upload
    openstack_public_key = "my-key.pem"

    # Path & filename of the SSH private key file
    openstack_ssh_private_key_file = "/root/terraform/configs/lattice-openstack/my-key.pem"

An help appreciated.
Dave

**Comment by matjohn2** [comment]
Hi dave.

That example is designed so you have 'zero touch' on the openstack environment. You literally give that variable a ssh public key in it's text format and terraform will then create the keypair for you in openstack.

For example.. If you create a local ssh public/private key pair on Mac/Linux with 'ssh-keygen' and following the instructions; the resulting contents of 'keyname.pub' can be copied out of the file and put into your openstack_public_key variable wrapped in quotes... Ie..

Openstack_public_key = "ssh-rsa eddb001749379adccbbcadedd... Etc etc"

Hope that helps, shout if not!

Matt

Sent from a mobile device.

On 4 May 2015, at 20:22, "David Botelho" <notifications@github.com<mailto:notifications@github.com>> wrote:


I keep getting the following error.

  *   Error creating OpenStack keypair: Expected HTTP response code [200] when accessing [POST http://vc2cmmka033430n.fmr.com:8774/v2/f2c27ae1651849e98f80b60231b752ef/os-keypairs], but got 400 instead {"badRequest": {"message": "Keypair data is invalid: failed to generate fingerprint", "code": 400}}

I'm thinking it has something to do with my config file entries below. I've created a keypair in my openstack tenant and saved the private key. That is the key used to ssh to each of the servers, so what am I missing?

# OpenStack Tenant Name
openstack_tenant_name = "my_dev"

# SSH Key Name
openstack_key_name = "my_keypair"

# SSH Public Key to Upload
openstack_public_key = "my-key.pem"

# Path & filename of the SSH private key file
openstack_ssh_private_key_file = "/root/terraform/configs/lattice-openstack/my-key.pem"


An help appreciated.
Dave

**Comment by davebotelho** [comment]
Hey Matt,

Thanks for that information.  I was able to get past that error and now it appears that lattice-coordinate and lattice-cell-0 vms are running.    The error I’m running into now seems to be that you are running an update on those VMs but they aren’t able to get out of our network because we have a proxy server that needs to be set on the VM before it would be able to do that.  Error message below:

module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec):   libmpc2 libmpfr4 libquadmath0
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec):   linux-libc-dev manpages-dev
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): 0 upgraded, 13 newly installed, 0 to remove and 0 not upgraded.
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Need to get 21.0 MB of archives.
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): After this operation, 55.0 MB of additional disk space will be used.
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): 0% [Working]
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Err http://archive.ubuntu.com/ubuntu/ precise/main libgomp1 amd64 4.6.3-1ubuntu5
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec):   Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): 0% [Connecting to archive.ubuntu.com]
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Err http://archive.ubuntu.com/ubuntu/ precise/main libmpfr4 amd64 3.1.0-3ubuntu2
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec):   Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): 0% [Connecting to archive.ubuntu.com]
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Err http://archive.ubuntu.com/ubuntu/ precise/main libquadmath0 amd64 4.6.3-1ubuntu5
/archive.ubuntu.com/ubuntu/pool/main/m/manpages/manpages-dev_3.35-0.1ubuntu1_all.deb  Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell: Error: 1 error(s) occurred:

* Script exited with non-zero exit status: 100
Error applying plan:

1 error(s) occurred:

* 1 error(s) occurred:

* 1 error(s) occurred:

* 1 error(s) occurred:

* Script exited with non-zero exit status: 100

Is there a way for me to set my internal proxy settings?  When I’ve deployed Cloudfoundry with BOSH in the past, I’ve had to crack the stemcell and add my proxy setttings manually into the image.  Will I have to do the same to the image I’m using?

Thanks,
Dave

**Comment by jbayer** [comment]
perhaps trying to set the proxy using these instructions would work?
https://help.ubuntu.com/community/AptGet/Howto#Setting_up_apt-get_to_use_a_http-proxy

**Comment by davebotelho** [comment]
I have the proxy set on my server.  I think the update is running on the lattice-coordinator vm – which does not have the proxy settings.  It was created form our Ubuntu image which does not have the apt proxy settings built into the image.

root@vc2cmmka033593n:~/terraform/configs/lattice-openstack# env | grep proxy

http_proxy=http://$MYPROXY:8000
https_proxy=http://$MYPROXY:8000

So my dilemma, do I crack the image and setup the proxy settings globaly or is there a proxy setting within the terraform openstack module.

Thanks,
Dave

**Comment by matjohn2** [comment]
I'm not somewhere I can hack on this so apologies for the words vs code response..

This shouldn't be that hard and you shouldn't have to hack the image.

Two options off the top of my head. The terraform scripts for building the instances can take cloud-init metadata environment settings... This could be used to put the proxy settings in the relevant configuration locations.

Alternatively; we could patch the setup scripts within lattice's repo. You'll notice these are unchanged across AWS/do/gce/openstack terraform providers.. So this would be another pull request to add support to the setup scripts across the board.

It's more generic to place in the metadata; it's easier to solve for all TF deployment possibilities to place in the Lattice install script.

James, thoughts?

Matt

On 5 May 2015, at 16:29, "David Botelho" <notifications@github.com<mailto:notifications@github.com>> wrote:

I have the proxy set on my server. I think the update is running on the lattice-coordinator vm – which does not have the proxy settings. It was created form our Ubuntu image which does not have the apt proxy settings built into the image.

root@vc2cmmka033593n:~/terraform/configs/lattice-openstack# env | grep proxy

http_proxy=http://$MYPROXY:8000
https_proxy=http://$MYPROXY:8000

So my dilemma, do I crack the image and setup the proxy settings globaly or is there a proxy setting within the terraform openstack module.

Thanks,
Dave

**Comment by davebotelho** [comment]
Hey Matt,

Thanks for that tip.  I edit the resources.tf file below with my metadata additions for proxy (was this the correct place?).

drwxr-xr-x 3 root root 4096 May  6 09:11 .
drwxr-xr-x 8 root root 4096 May  5 08:42 ..
drwxr-xr-x 2 root root 4096 May  5 08:42 example
-rw-r--r-- 1 root root  252 May  5 08:42 outputs.tf
-rw-r--r-- 1 root root  216 May  5 08:42 provider.tf
-rw-r--r-- 1 root root 3883 May  5 08:42 README.md
-rw-r--r-- 1 root root 7128 May  6 09:11 resources.tf
-rw-r--r-- 1 root root 2281 May  5 08:42 variables.tf
root@vc2cmmka033593n:~/terraform/configs/lattice-openstack/.terraform/modules/4fa7e35a48e762327e915bfa31f0adbb/terraform/openstack#

When I re-ran the terraform apply I did notice the meta data variables being set (as seen below):

  image_id:              "" => "<computed>"
  image_name:            "" => "ubuntu1204_nonfid_cloud_v2"
  key_pair:              "" => "terraform_keypair"
  metadata.#:            "" => "3"
  metadata.http_proxy:   "" => "http://MYPROXY:8000"
  metadata.https_proxy:  "" => "http://MYPROXY:8000"
  metadata.lattice-role: "" => "coordinator"
  name:                  "" => "lattice-coordinator"
  network.#:             "" => "1"
  network.0.fixed_ip_v4: "" => "<computed>"
  network.0.fixed_ip_v6: "" => "<computed>"
  network.0.mac:         "" => "<computed>"
  network.0.name:        "" => "<computed>"

BUT, I did get an error when it was trying to download:

module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): After this operation, 55.0 MB of additional disk space will be used.
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): 0% [Working]
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Err http://archive.ubuntu.com/ubuntu/ precise/main libgomp1 amd64 4.6.3-1ubuntu5
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec):   Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): 0% [Working]
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Err http://archive.ubuntu.com/ubuntu/ precise/main libmpfr4 amd64 3.1.0-3ubuntu2
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec):   Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): 0% [Working]
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Err http://archive.ubuntu.com/ubuntu/ precise/main libquadmath0 amd64 4.6.3-1ubuntu5
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec):   Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): 0% [Working]
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Err http://archive.ubuntu.com/ubuntu/ precise/main libmpc2 amd64 0.9-4
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec):   Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): 0% [Working]
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Err http://archive.ubuntu.com/ubuntu/ precise-updates/main binutils amd64 2.22-6ubuntu1.1
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec):   Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): 0% [Working]
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Err http://archive.ubuntu.com/ubuntu/ precise/main cpp-4.6 amd64 4.6.3-1ubuntu5
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec):   Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): 0% [Connecting to archive.ubuntu.com]
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Err http://archive.ubuntu.com/ubuntu/ precise/main cpp amd64 4:4.6.3-1ubuntu5
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec):   Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): 0% [Working]
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Err http://archive.ubuntu.com/ubuntu/ precise/main gcc-4.6 amd64 4.6.3-1ubuntu5
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec):   Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): 0% [Connecting to security.ubuntu.com]
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Err http://security.ubuntu.com/ubuntu/ precise-security/main binutils amd64 2.22-6ubuntu1.1
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec):   Temporary failure resolving 'security.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): 0% [Connecting to archive.ubuntu.com]
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Err http://archive.ubuntu.com/ubuntu/ precise/main gcc amd64 4:4.6.3-1ubuntu5
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec):   Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): 0% [Working]
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Err http://archive.ubuntu.com/ubuntu/ precise-updates/main libc-dev-bin amd64 2.15-0ubuntu10.5
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec):   Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): 0% [Working]
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Err http://security.ubuntu.com/ubuntu/ precise-security/main libc-dev-bin amd64 2.15-0ubuntu10.5
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec):   Temporary failure resolving 'security.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): 0% [Connecting to archive.ubuntu.com]
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Err http://archive.ubuntu.com/ubuntu/ precise-updates/main linux-libc-dev amd64 3.2.0-60.91
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec):   Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): 0% [Working]
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Err http://security.ubuntu.com/ubuntu/ precise-security/main linux-libc-dev amd64 3.2.0-60.91
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec):   Temporary failure resolving 'security.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): 0% [Connecting to archive.ubuntu.com]
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Err http://archive.ubuntu.com/ubuntu/ precise-updates/main libc6-dev amd64 2.15-0ubuntu10.5
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec):   Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): 0% [Working]
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Err http://security.ubuntu.com/ubuntu/ precise-security/main libc6-dev amd64 2.15-0ubuntu10.5
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec):   Temporary failure resolving 'security.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): 0% [Connecting to archive.ubuntu.com]
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Err http://archive.ubuntu.com/ubuntu/ precise/main manpages-dev all 3.35-0.1ubuntu1
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec):   Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): 0% [Working]
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/g/gcc-4.6/libgomp1_4.6.3-1ubuntu5_amd64.deb  Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/m/mpfr4/libmpfr4_3.1.0-3ubuntu2_amd64.deb  Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/g/gcc-4.6/libquadmath0_4.6.3-1ubuntu5_amd64.deb  Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/m/mpclib/libmpc2_0.9-4_amd64.deb  Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/b/binutils/binutils_2.22-6ubuntu1.1_amd64.deb  Temporary failure resolving 'security.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/g/gcc-4.6/cpp-4.6_4.6.3-1ubuntu5_amd64.deb  Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/g/gcc-defaults/cpp_4.6.3-1ubuntu5_amd64.deb  Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/g/gcc-4.6/gcc-4.6_4.6.3-1ubuntu5_amd64.deb  Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/g/gcc-defaults/gcc_4.6.3-1ubuntu5_amd64.deb  Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/e/eglibc/libc-dev-bin_2.15-0ubuntu10.5_amd64.deb  Temporary failure resolving 'security.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/l/linux/linux-libc-dev_3.2.0-60.91_amd64.deb  Temporary failure resolving 'security.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/e/eglibc/libc6-dev_2.15-0ubuntu10.5_amd64.deb  Temporary failure resolving 'security.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/m/manpages/manpages-dev_3.35-0.1ubuntu1_all.deb  Temporary failure resolving 'archive.ubuntu.com'
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell: Error: 1 error(s) occurred:

* Script exited with non-zero exit status: 100
Error applying plan:

1 error(s) occurred:

* 1 error(s) occurred:

* 1 error(s) occurred:

* 1 error(s) occurred:

* Script exited with non-zero exit status: 100

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.

**Comment by jbayer** [comment]
thanks @matjohn2 

@menicosia is the PM on the project, and should be able to consult with @davidwadden and @kaayell to determine the best approach to enabling proxies and sharing that info and whether they have short-terms plans to do that or perhaps a PR to do that would be the faster way to deliver that capability.

**Comment by davebotelho** [comment]
I had to edit all your scripts and add my proxy info but I got it to deploy.  

module.lattice-openstack.openstack_compute_instance_v2.lattice-cell (remote-exec): Finished Installing Lattice Cell (v0.2.3)!
module.lattice-openstack.openstack_compute_instance_v2.lattice-cell: Creation complete

Apply complete! Resources: 1 added, 3 changed, 0 destroyed.

The state of your infrastructure has been saved to the path
below. This state is required to modify and destroy your
infrastructure, so keep it safe. To inspect the complete state
use the `terraform show` command.

State path: terraform.tfstate

Outputs:

  http_proxy       = http://http.proxy.fmr.com:8000
  https_proxy      = http://http.proxy.fmr.com:8000
  lattice_password = pass
  lattice_target   = 10.207.138.113.xip.io
  lattice_username = user

Cool.  Maybe there could be something added to add both http and https proxy info to the terraform config file.  I'll let you know if it works!
Thanks,
Dave

**Comment by davebotelho** [comment]
Hi Matt,

Something new this morning...I did a terraform get -update and now I get this error:

root@docker1:~/terraform/configs/lattice-openstack# terraform plan
There are warnings and/or errors related to your configuration. Please
fix these before continuing.

Errors:

  * module root: module lattice-openstack: auth_url is not a valid parameter
  * provider.openstack: "auth_url": required field is not set

I'm already providing the keystone_uri in my lattice config.

   # URI of Keystone authentication agent
    openstack_keystone_uri = "http://vc2cmmka033430n.fmr.com:5000/v2.0"

Why am I getting this now?

Thanks,
Dave

**Comment by ajlanghorn** [comment]
Daft question, perhaps, @davebotelho, but are you actually passing a value of `auth_url` to Terraform somehow (eg at the prompt, in a `.tfvars` etc.)?

**Comment by davebotelho** [comment]
In the previous module version (prior to doing the update), it didn't ask me for this auth_url.  When I received it today, I was confused because the only auth_url I could think of was keystone which I am already passing.  I'm confused as to what auth_url it is asking for now.  Your question is my question.

**Comment by menicosia** [comment]
Hi @matjohn2,

Additional feedback on the PR: One of SA's tested this for me, and found the following:
```
I tested it against a Mirantis Express 6 environment and it works, with one exception: the user and password don't get passed to the receptor. It's always "user" and "pass".
It would be good to be able to specify the UUIDs of pre-existing components so Terraform won't just fail if they cannot be provisioned. I tested in DreamCompute and it failed due to the restrictive nature of the public environment.
```

There's a lot of interest in this PR, so we are eager to incorporate it in the next of Lattice, please let me know if we can help.

--
  Marco Nicosia
  Product Manager
  Pivotal Software, Inc.

**Comment by matjohn2** [comment]
Hey @menicosia.

I can't reproduce your SA's issue with the username and password.
The defaults are user/pass, but adding the lattice_username and lattice_password variables into the `lattice.openstack.tf` file allows me to specify different ones and these then work correctly with the `$ ltc target` command.

```
matt@MATJOHN2-M-401J lattice-openstack $ ltc ta 10.1.2.34.xip.io
Username: matt
Password: 
Api Location Set
matt@MATJOHN2-M-401J lattice-openstack $ ltc test
Running Suite: Lattice Integration Tests
========================================
Random Seed: 1431544195
Will run 2 of 2 specs

• [SLOW TEST:130.839 seconds]
Lattice
/root/go/src/github.com/cloudfoundry-incubator/lattice/ltc/integration_test/integration_test_runner.go:135
  when desiring a docker-based LRP
  /root/go/src/github.com/cloudfoundry-incubator/lattice/ltc/integration_test/integration_test_runner.go:134
    eventually runs a docker app
    /root/go/src/github.com/cloudfoundry-incubator/lattice/ltc/integration_test/integration_test_runner.go:127
------------------------------
•Ran 2 of 2 Specs in 136.328 secondsSUCCESS! -- 2 Passed | 0 Failed | 0 Pending | 0 Skipped ```

**Comment by menicosia** [comment]
@matjohn2 we merged this PR regardless of the lingering issues. OpenStack is just a devil to accept, there's a lot of variability in OpenStack setups.

I really wanted to get this into v0.2.5, so we went ahead. We'll mark this is "community contributed," and hope that those who wish to push forward on OpenStack will continue to issue PRs. If we get better capabilities to accept internally, we'll put what we can behind validating that PRs are functional.

--
  Marco Nicosia
  Product Manager
  Pivotal Software, Inc.

**Pull request merged**